docs/dev/wasm.md

```diff
@@ -50,75 +50,69 @@ and passing for each parameter a 64-bit value, of which top 32 bits are its size
 }
 ```
 
-## Support for NULL values
+## Parameter values
 
 Native WebAssembly types can only be represented directly if the function does not operate on NULL values. Fortunately, user-defined functions
 explicitly specify whether they accept NULL or not.
 
-If the function is specified not to accept NULL, all parameters are represented
-as in the description above.
+If the function is specified not to accept NULL, the parameter types natively supported by WASM are represented by their values (types shorter than 4 bytes are extended to 4 bytes),
+and all remaining types are represented as in the description above.
 
 If the function is specified to accept NULL, each parameter should be represented in WebAssembly by 
 size and pointer to its serialized form in memory, as in the paragraph above.
 
-the important distinction is that size equal to -1 (minus one)indicates that the value is NULL and should not be parsed.
+The important distinction is that size equal to -1 (minus one or i32::MAX) indicates that the value is NULL and should not be parsed.
 
 ## Return values
 
-NOTE: ABI for return values is experimental and subject to change. It can (and should) be redesigned
-after implementing helper libraries for a few popular languages (including C++, C, Rust).
+NOTE: CQL syntax extensions, and new helper libraries may be deployed together with new ABI versions,
+the description below only refers to ABI versions 1 and 2.
 
-Natively supported types are returned as is. All the other types are returned via memory, similarly to the way
-they are passed as parameters: the wasm function should return a single 64-bit integer, representing the size and the offset of a serialized form of the returned value
+In `RETURNS NULL ON NULL INPUT` UDFs, natively supported types are returned as is. All the other types are returned via memory,
+similarly to the way they are passed as parameters: the wasm function should return a single 64-bit integer, representing
+the size and the offset of a serialized form of the returned value
```

Comment:
Let's add information about memory management. Who is responsible to free the memory, both for inputs and return values? How is memory freed?

Reply:
We already have this: https://github.com/scylladb/scylladb/blob/64102780fe570b53617594900eba1e229e537a3c/docs/dev/wasm.md?plain=1#L21-L29, is there anything to add?

Reply:
I forgot it. But yes, we should document the return value too.

One trick: suppose we have an aggregation function f(accumulator, arg) -> accumulator. The user is allowed to move the first argument into the return value (this moves the responsibility of freeing to the caller). We can then take advantage of this and *not* call _scylla_free on the return value; instead, we pass the return value back the next time we aggregate. This saves us the ffi overhead of extracting the accumulator and copying it back again, if we can reuse the function.